### PR TITLE
[stdlib] Unicode 9 and Thread Local Storage (again)

### PR DESCRIPTION
**Pull request by milseman**
TL;DR Add support for Unicode 9 grapheme breaking (i.e. emoji) by using ICU. Work around ICU efficiency issues with thread-local storage.

NOTE: This is an update on a previous PR that had to be reverted due to the unfortunate dependencies introduced. This untangles some of them by forward declaring the functions we use on Apple platforms, which will typically not be building corelibs-foundation and/or their own version of ICU. What follows is from the original PR.

---

Introduce shims for using UBreakIterators from ICU. Also introduce
shims for using thread local storage via pthreads.
    
We will be relying on ICU and UBreakIterators for grapheme
breaking. But, UBreakIterators are very expensive to create,
especially for the way we do grapheme breaking, which is relatively
stateless. Thus, we will stash one or more into thread local storage
and reset it as needed.
    
---

Introduces a _ThreadLocalStorage struct to hold thread-local, but
global resources. Set it up to host a UBreakIterator and a cache key
for resetting text.
    
UBreakIterators are extremely expensive to create on the fly, so we
store one for each thread. UBreakIterators are also expensive to bind
to new text, so we cache the text it's currently bound to in order to
help avoid it.
    
The struct can be expanded with more functionality in the future, but
the standard library should only ever use a single key, and thus
everything should go on this struct. The _ThreadLocalStorage struct is
not meant to be copyable, creatable (by anyone else except the
once-per-thread initialize routine), and should accessed through the
pointers it provides.
    
Future immediate directions could include cashing multiple
UBreakIterators (e.g. avoid a text reset for mutual character
iteration patterns, etc).
    
Test added in test/stdlib/ThreadLocalStorage.swift.

---

Use UBreakIterators to perform grapheme breaking. This gives Unicode 9
grapheme breaking (e.g. family emoji) and provides a means to upgrade
to future versions. It also serves as a model for how to serve up
other advanced functionality in ICU to users.
    
This has tricky performance implications. Some things are faster and a
number of cases are slower. But, careful use of ICU can help mitigate
and amortize these costs. In conjunction with more early detection of
fast paths, overall grapheme breaking for the average user should be
much faster than in Swift 3.
    
NOTE: This is incomplete. It currently falls back on the legacy tries
for some bridged strings. There are many potential directions for a
general solution, but for now we'll be interatively adding support for
more and more special cases.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by shahmishal** [comment]
@swift-ci please test

**Comment by shahmishal** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@shahmishal have you seen this source of failure before?
```
00:22:19 dyld: Library not loaded: libFoundationTestsShared.dylib
```

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

**Review by airspeedswift — APPROVED** [review]
LGTM

File: stdlib/public/core/ThreadLocalStorage.swift [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    -> UnsafeMutablePointer<_ThreadLocalStorage>
+  {
+    let tlsRawPtr = _swift_stdlib_pthread_getspecific(_tlsKey)
+    if _fastPath(tlsRawPtr != nil) {
+      return tlsRawPtr._unsafelyUnwrappedUnchecked.assumingMemoryBound(
+        to: _ThreadLocalStorage.self)
+    }
+
+    return _initializeThreadLocalStorage()
+  }
```
Comment (atrick) [inline stdlib/public/core/ThreadLocalStorage.swift]: I'm sure you already know... but if you move getPointer into the C runtime you can use clang's TLS support when available. That will remove all of the pthread call overhead.

Comment (milseman) [inline stdlib/public/core/ThreadLocalStorage.swift]: No I do not know, thanks! Could you file an SR for that?

Comment (atrick) [inline stdlib/public/core/ThreadLocalStorage.swift]: https://bugs.swift.org/browse/SR-4915

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

**Comment by milseman** [comment]
@swift-ci please clean test OS X Platform

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

(not sure if I triggered that before the revert)

**Comment by milseman** [comment]
@swift-ci please test OS X Platform

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test OS X platform
